Apps/Patient/src/Services/PatientDataService.cs

```diff
@@ -49,13 +62,29 @@ public async Task<PatientDataResponse> Query(PatientDataQuery query, Cancellatio
         public async Task<PatientFileResponse?> Query(PatientFileQuery query, CancellationToken ct)
         {
             Guid pid = await this.ResolvePidFromHdid(query.Hdid).ConfigureAwait(true);
-            PatientFile? file = (await this.patientDataRepository.Query(new PatientDataAccess.PatientFileQuery(pid, query.FileId), ct).ConfigureAwait(true)).Items.OfType<PatientFile>().FirstOrDefault();
+            PatientFile? file = (await this.patientDataRepository.Query(new PatientDataAccess.PatientFileQuery(pid, query.FileId), ct).ConfigureAwait(true)).Items.OfType<PatientFile>()
+                .FirstOrDefault();
 
             return file != null
                 ? new PatientFileResponse(file.Content, file.ContentType)
                 : null;
         }
 
+        private async Task<bool> CanAccessDataSourceAsync(string hdid, IList<PatientDataType> patientDataTypes)
+        {
+            for (int i = patientDataTypes.Count - 1; i >= 0; i--)
+            {
+                DataSource dataSource = this.mapper.Map<DataSource>(patientDataTypes[i]);
+                if (!await this.patientRepository.CanAccessDataSourceAsync(hdid, dataSource).ConfigureAwait(true))
+                {
+                    // Remove patient data type from query if access has been blocked.
+                    patientDataTypes.RemoveAt(i);
+                }
+            }
+
+            return patientDataTypes.Count != 0;
+        }
```

Comment:
I like Declan's idea, but I think the async stuff might be too fancy (if trying to avoid the the blocking call to .Result).

Your method is doing 2 things - filtering the data types and counting them - when it could focus on just the filtering.

```
private static async Task<IList<PatientDataType>> GetUnblockedPatientDataTypesAsync(string hdid, IEnumerable<PatientDataType> patientDataTypes)
{
    List<PatientDataType> unblockedPatientDataTypes = new();
    foreach (PatientDataType patientDataType in patientDataTypes)
    {
        DataSource dataSource = this.mapper.Map<DataSource>(patientDataType);
        if (await CanAccessDataSourceAsync(hdid, dataSource).ConfigureAwait(true))
        {
            unblockedPatientDataTypes.Add(patientDataType);
        }
    }

    return unblockedPatientDataTypes;
}
```

Usage:

```
public async Task<PatientDataResponse> Query(PatientDataQuery query, CancellationToken ct)
{
    IList<PatientDataType> unblockedPatientDataTypes = await GetUnblockedPatientDataTypesAsync(query.PatientDataTypes).ConfigureAwait(true));
    if (!unblockedPatientDataTypes.Any())
    {
        return new PatientDataResponse(Enumerable.Empty<PatientData>());
    }
```

Reply:
Changed